End-to-end acknowledgements in data prepper - blog post (#1682)
opensearch-project/project-website: _posts/2023-06-16-End-to-end-acknowledgements-in-Data-Prepper.md

Several challenges arise with data durability. For example, Data Prepper may exhaust memory or other system resources, or the hardware running Data Prepper may fail. The maintainers of Data Prepper and their teams have observed that the most common hindrance to data durability is the sink. If the OpenSearch cluster is unable to receive data either due to temporary stress on the cluster or if it is under-scaled, then Data Prepper cannot send data to the destination sink.

We have also considered the sources of data. In particular, we have noticed that the Amazon S3 source is already reading data from a highly durable store. Keeping these observations in mind we realized that if Data Prepper fails to deliver data to OpenSearch, it can retry reading from S3. We just need to know when the data is written before deleting the SQS message that notifies Data Prepper of an available S3 object to consume.

## Our solution: End-to-end acknowledgements

> Data Prepper provides data durability through the use of end-to-end acknowledgements. When a Data Prepper source is configured to use end-to-end acknowledgements, the source is notified only when the data is successfully delivered to the sink. If the source receives end-to-end acknowledgments, it can take appropriate actions such as removing SQS messages that have been successfully delivered to the sink or incrementing the commit offset in case of Kafka source. If end-to-end acknowledgements are not received, the source may retry operation or notify the external source of the failure. The following picture illustrates the control flow when a Data Prepper source is configured to use end-to-end acknowledgements.

Reviewer: ```suggestion
When a Data Prepper source is configured to use end-to-end acknowledgements, the source is notified only when the data is successfully delivered to the sink. If the source receives end-to-end acknowledgments, it can take appropriate actions, such as removing SQS messages that have been successfully delivered to the sink or incrementing the commit offset in the case of a Kafka source. If end-to-end acknowledgements are not received, the source may retry the operation or notify the external source of the failure. The following diagram shows the control flow of a Data Prepper source configured to use end-to-end acknowledgements.
```

<img src="/assets/media/blog-images/2023-06-16-end-to-end-acknowledgements/end-to-end-acknowledgements.png" alt="End-to-end acknowledgements diagram"/>{: .img-fluid}

> 1. Data Prepper source receives a batch of records from an external source (such as Amazon S3).

Reviewer: ```suggestion
1. The Data Prepper source receives a batch of records from an external source, such as Amazon S3.
```

> 2. When a Data Prepper source is configured with end-to-end acknowledgements, the source creates an AcknowledgementSet along with a callback function for each batch of records received from the external source. The Source converts the records to Data Prepper events

Reviewer 2: Should `AcknowledgementSet` be in code format?  Furthermore, where can I find the AcknowledgementSet code?

3. Each event in the batch is added to the AcknowledgementSet, and a reference to the AcknowledgementSet is kept in the event as a handle

> 4. Events are then passed to the ingestion pipeline, which consists of multiple processors that transform, filter, enrich the data in the event

Reviewer: ```suggestion
4. Events are then passed to the ingestion pipeline, which consists of [processors](https://opensearch.org/docs/latest/data-prepper/pipelines/configuration/processors/processors/) that transform, filter, or enrich the data in the event.
```

> 5. If an event is dropped in the processors, the AcknowledgementSet is notified of the completion of the event when the event handle is released

Reviewer: ```suggestion
5. If an event is dropped in the processors, the AcknowledgementSet is notified of the event completion when the event handle is released.
```

> 6. All the events that are not dropped are sent to the Data Prepper sink

Reviewer: ```suggestion
6. Events that are not dropped are sent to the Data Prepper sink.
```

> 7. The Data Prepper sink sends the events to the external sink (such as OpenSearch Cluster)

Reviewer: ```suggestion
7. The Data Prepper sink sends the events to the external sink, for example, the OpenSearch cluster.
```

8. When events are successfully sent to the external sink, the AcknowledgementSet is notified of the event's completion when the event handles are released.
9. When the AcknowledgementSet has no pending events, it invokes the callback function

> 10. The callback function notifies the external source about the successful delivery of the batch of records

Reviewer: ```suggestion
10. The callback function notifies the external source of the successful delivery of the records batch.
```

> If all events are not successfully delivered to the external sinks, AcknowledgementSet will retain some events that have not been released. Each AcknowledgementSet has an expiry time, and if all events are not released before the expiry time, the AcknowledgementSet is freed, and external source is not notified of the delivery of the records. This situation may trigger Data Prepper source or the external source to take corrective action, such as retrying the ingestion of the records again.

Reviewer: ```suggestion
If all events are not successfully delivered to the external sinks, AcknowledgementSet retains some events that have not been released. Each AcknowledgementSet has an expiry time. If all events are not released before the expiry time, the AcknowledgementSet is freed and the external source is not notified of the delivery of the records. This situation may trigger the Data Prepper source or external source to take corrective action, such as retrying the records ingestion.
```

> Data Prepper supports multiple pipelines and events can be routed to different pipelines based on the configuration and the data within the events. Additionally, each event may be routed to multiple pipelines that eventually send data to different sinks. However, in some cases, multiple copies of the same event are not created when it is routed to multiple pipelines/sinks. This poses a challenge when tracking acknowledgements because multiple acknowledgements are required for the same event. To address this issue, reference counts are maintained for events that are routed to multiple pipelines. When an event handle is released (either in the step 5 or step 8), the reference count is decremented. The event is considered completed and the handle is removed from the AcknowledgementSet only when the reference count reaches zero, indicating the successful delivery of the event to all sinks. This approach ensures accurate tracking of acknowledgements for events routed to multiple pipelines.

Reviewer: ```suggestion
Data Prepper supports multiple pipelines, and events can be routed to different pipelines based on the configuration and the data within the events. Each event also may be routed to multiple pipelines that send data to different sinks. In some cases, multiple copies of the same event are not created when the event is routed to multiple pipelines or sinks. This can be a challenge when tracking acknowledgements because multiple acknowledgements are required for the same event. Thus, reference counts are maintained for events that are routed to multiple pipelines. When an event handle is released (either in step 5 or step 8), the reference count is decremented. The event is considered complete and the handle is removed from the AcknowledgementSet when the reference count reaches zero, which indicates successful delivery of the event to all sinks. This approach provides for accurate tracking of acknowledgements for events routed to multiple pipelines.
```

> Data Prepper also provides support for sending negative acknowledgements to indicate explicit failure. The callback function can examine the acknowledgement status, whether it is positive or negative, and take appropriate action based on that information.

Reviewer: ```suggestion
Data Prepper also can send negative acknowledgements to indicate explicit failure. The callback function can examine the acknowledgement status, determine whether it is positive or negative, and take appropriate action based on that information.
```

> In certain scenarios, when Data Prepper sinks are configured with Dead Letter Queue (DLQ), events that cannot be delivered to the external sink (such as OpenSearch) are written to the DLQ. When end-to-end acknowledgements are enabled, successfully writing the events to the DLQ (after failing to deliver them to the external sink) is considered a successful completion of event delivery. In this case, a positive acknowledgement is delivered to the acknowledgment set, indicating successful processing and handling of the event.

Reviewer: ```suggestion
In certain scenarios, when Data Prepper sinks are configured with Dead Letter Queue (DLQ), events that cannot be delivered to the external sink, such as OpenSearch, are written to DLQ. When end-to-end acknowledgements are enabled, successfully writing the events to DLQ (after failing to deliver them to the external sink) is considered a successful completion of event delivery. In this case, a positive acknowledgement is delivered to the AcknowledgmentSet, indicating successful event processing and handling.
```

## Moving forward: Conclusion and next steps

> End-to-end acknowledgments offer robust data durability when utilizing Data Prepper for data processing and ingestion. However, there are a couple of important considerations regarding this feature:

Reviewer: ```suggestion
End-to-end acknowledgments offer robust data durability when using Data Prepper for data processing and ingestion. Some considerations, however, are necessary:
```

---
layout: post
title:  "End-to-end acknowledgements in Data Prepper"
authors:
- kkondaka
- dlv
date: 2023-06-16 10:00:00 -0700
categories:
  - technical-post

meta_keywords: Data Prepper E2E, Data Prepper sink, Data Prepper pipeline as the sink
meta_description: Learn how Data Prepper provides data durability through the use of end to end acknowledgements by delivering data to the sink before notifying the source.

---

> ## The need: Improving durability

Reviewer 2: Lets remove this heading. It looks odd on the rendered site and the same idea is conveyed with the line "To solve challenges with data durability, Data Prepper now has capability to use end-to-end acknowledgements."

> ## The need: Improving durability

Reviewer 2: ```suggestion

```

> [Data Prepper](https://opensearch.org/docs/latest/data-prepper/index/) is an open-source data collector used to ingest data into OpenSearch clusters. To solve challenges with data durability, Data Prepper now has capability to use end-to-end acknowledgements.

Reviewer 2: Lets introduce what challenge E2E solves in the intro. 

```suggestion
[Data Prepper](https://opensearch.org/docs/latest/data-prepper/index/) is an open-source data collector used to ingest data into OpenSearch clusters. While Data Prepper has an in-memory buffer that allows for fast throughput, users also want improved data durability, particularly, confirmation that the data received by Data Prepper reaches the desired sink.  To solve challenges with data durability, Data Prepper now has capability to use end-to-end acknowledgements.
```

> 

Reviewer 2: Add the previous heading here:

```suggestion
## The need: Improving durability

```

> While Data Prepper has an in-memory buffer that allows for fast throughput, users also want improved data durability, particularly, confirmation that the data received by Data Prepper reaches the desired sink. Data Prepper maintainers and their teams have observed that a common challenge to data durability is the sink. For example, if the OpenSearch cluster cannot receive data due to temporary stress on the cluster or underscaling, then Data Prepper cannot send data to the destination sink. Further, Data Prepper may exhaust memory or other system resources or the hardware running Data Prepper may fail.

Reviewer: ```suggestion
While Data Prepper has an in-memory buffer that allows for fast throughput, users also want improved data durability, particularly, confirmation that the data received by Data Prepper reaches the desired sink. Data Prepper maintainers and their teams have observed that a common challenge to data durability is the sink. For example, if the OpenSearch cluster cannot receive data because of temporary stress on the cluster or underscaling, then Data Prepper cannot send data to the destination sink. Further, Data Prepper may exhaust memory or other system resources or the hardware running Data Prepper may fail.
```

> While Data Prepper has an in-memory buffer that allows for fast throughput, users also want improved data durability, particularly, confirmation that the data received by Data Prepper reaches the desired sink. Data Prepper maintainers and their teams have observed that a common challenge to data durability is the sink. For example, if the OpenSearch cluster cannot receive data because of temporary stress on the cluster or underscaling, then Data Prepper cannot send data to the destination sink. Further, Data Prepper may exhaust memory or other system resources or the hardware running Data Prepper may fail.

Reviewer 2: The durability of the sink's cluster? Since we're referring to the OpenSearch cluster as the sink destination, correct?

Also, lets add this first sentence to the intro, per the preceding comment.

```suggestion
Data Prepper maintainers and their teams have observed that a common challenge to data durability is the durability sink's cluster. For example, if the OpenSearch cluster cannot receive data because of temporary stress on the cluster or underscaling, then Data Prepper cannot send data to the destination sink. Further, Data Prepper may exhaust memory or other system resources or the hardware running Data Prepper may fail.
```

> We have also considered the sources of data. In particular, we have noticed that the Amazon S3 source is already reading data from a highly durable store. Keeping these observations in mind we realized that if Data Prepper fails to deliver data to OpenSearch, it can retry reading from S3. We just need to know when the data is written before deleting the SQS message that notifies Data Prepper of an available S3 object to consume.

Reviewer 2: Echoing Melissa's suggestion with a couple additional clarifying adjectives.

```suggestion
To solve this challenge with data durability, we must consider the data sources themselves. For example, the Amazon S3 source can read data from a highly durable store. When observing this, we realized that if Data Prepper fails to deliver data to OpenSearch, Data Prepper can retry reading data from Amazon S3. This, however, requires knowing when the data is written before deleting the SQS message that notifies Data Prepper of an available S3 object to consume.
```
